Adds provider package documentation in installation.rst (#12203)
mistercrunch/Airflow: docs/installation.rst

    CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"
    pip install "apache-airflow[postgres,google]==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"

Most of the extras are linked to a corresponding providers package. For example "amazon" extra

> has a corresponding 'apache-airflow-providers-amazon' providers package to be installed. When you install

Reviewer: ```suggestion
has a corresponding ``apache-airflow-providers-amazon`` providers package to be installed. When you install
```

| winrm               | ``pip install 'apache-airflow[microsoft.winrm]'``   | WinRM hooks and operators                                            |
+---------------------+-----------------------------------------------------+----------------------------------------------------------------------+

Provider packages
.................

> Provider packages context

Reviewer 2: My perception could be wrong, but a big benefit I see to this feature is how "dependency-light" it makes the Airflow base. Since all of these providers (which have their own sets of complex dependencies) are no longer included in the base, a user is less likely to run into one-off dependency conflicts when trying to install additional libraries. I wonder if that's worth calling out in the context here? Curious to hear your thoughts @potiuk.

Author: Well, it actually was this way already. 

Unless you installed extras, the number of dependencies for the core airlfow was rather small and this has not changed with "provider packages" - we have basically the same set of requirements as before. If you did 'pip install apache-airflow' - the same dependencies are pulled in 1.10 and 2.0. And when you do 'pip install apache-airflow[google]'  - it pulls in the same set of dependencies in 1.10 and 2.0 (but in 2.0 those will be transitive dependencies for google provider, not the extra dependencies). The only difference was that if you have not installed 'google' extra before, the google operators were there in 'airflow.contrib', but they could not be imported because they missed dependencies.

So technically speaking - 2.0 is not any "lighter" in terms of dependencies. It behaves very much the same as it did before.

Maybe we can add some wording around it but literally, the only difference was that previously the classes were there (unusable) and now they are gone if you don't explicitly add 'extra' or install provider package manually. All dependencies remain the same. So for me this is not a differentiating feature. But maybe it is indeed worth spelling out (but I am not sure how :))

Reviewer 2: Ahh that makes sense- I didn't realize this was how it worked pre-2.0! Thanks for clearing it up, probably ok to leave as is for now then.

constraining the Airflow version used, it will be included as limitation of dependencies in the provider
package.

Some of the providers have cross-provider dependencies as well. Those are not required dependencies, they
might simply enabled certain features (for example transfer operators often create dependency between
different providers. Again, the general approach her is that the providers are backwards compatible,
including cross-dependencies. Any kind of breaking changes and requirements on particular versions of other
provider packages are automatically documented in the release notes of every provider.

.. note::
    We also provide ``apache-airflow-backport-providers`` packages that can be installed for Airflow 1.10.
    Those are the same providers as for 2.0 but automatically back-ported to work for Airflow 1.10. Those
    backport providers are going to be updated and released for 3 months after Apache Airflow 2.0 release.

Provider packages functionality
-------------------------------

Separate provider packages provide the possibilities that were not available in 1.10:

> 1. You can upgrade to latest version of particular providers without the need of Apache Airflow core upgrade

Reviewer 2: Missing a period here.

   to also upgrade provider packages at the same time?

A. It depends on your use case. If you have automated or semi-automated verification of your installation,
   that you can run a new version of Airflow including all provider packages, then definitely go for it.
   If you rely more on manual testing, it is advised that you upgrade in stages. Depending on your choice
   you can either upgrade all used provider packages first, and then upgrade Airflow Core or the other way
   round. The first approach - when you first upgrade all providers is probably safer, as you can do it
   incrementally, step-by-step replacing provider by provider in your environment.

Q. I have an Airflow version (1.10.12) running and it is stable. However, because of a Cloud provider change,
   I would like to upgrade the provider package. If I don't need to upgrade the Airflow version anymore,
   how do I know that this provider version is compatible with my Airflow version?

A. Backport Provider Packages (those are needed in 1.10.* Airflow series) are going to be released for
   3 months after the release. We will stop releasing new updates to the backport providers afterwards.

>    You will ba able to continue using the provider packages that you already use and unless you need to

Reviewer 2: Small typo here- `be` :)

>    get some new release of the provider that is only released in master, there is no need to upgrade

Reviewer 2: What would be a case in which this happens (ie. a provider is only released in master)? If the BaseOperator changes or something?

Author: Clarified. Not really. We expect some future changes to providers. A good example might be Google Python client 2.0.*. Google released 2.0 python clients recently for most of their services. Those are not backwards compatible. Most of the google providers for 2.0 will eventually migrate to those libraries and it's rather unlikely this will happen before the 3 months pass after Airflow 2.0, so those providers based on new clients will only be available for Airflow 2.0 (when the providers migrate to it) because we will not release backport providers then.

providers that have corresponding extras, the provider packages (latest version from PyPI) are installed
automatically when Airflow is installed with the extra.

Providers are released and versioned separately from the Airflow releases. We are following the
`Semver <https://semver.org/>`_ versioning scheme for the packages. Some versions of the provider
packages might depend on particular versions of Airflow, but the general approach we have is that unless
there is a good reason, new version of providers should work with recent versions of Airflow 2.x. Details
will vary per-provider and if there is a limitation for particular version of particular provider,
constraining the Airflow version used, it will be included as limitation of dependencies in the provider
package.

Some of the providers have cross-provider dependencies as well. Those are not required dependencies, they
might simply enabled certain features (for example transfer operators often create dependency between
different providers. Again, the general approach her is that the providers are backwards compatible,
including cross-dependencies. Any kind of breaking changes and requirements on particular versions of other

> provider packages are automatically documented in the release notes of every provider.

Reviewer: ```suggestion
Some of the providers have cross-provider dependencies as well. Those are not required dependencies, they
might simply enabled certain features (for example transfer operators often create dependency between
different providers. Again, the general approach her is that the providers are backwards compatible,
including cross-dependencies. Any kind of breaking changes and requirements on particular versions of other
provider packages are automatically documented in the release notes of every provider.
```
```suggestion
Some of the providers have cross-provider dependencies as well. Those are not required dependencies, they
might simply enable certain features (for example transfer operators often create a dependency between
different providers. Again, the general approach here is that the providers are backwards compatible,
including cross-dependencies. Any kind of breaking changes and requirements on particular versions of other
provider packages are automatically documented in the release notes of every provider.
```

.. note::
    We also provide ``apache-airflow-backport-providers`` packages that can be installed for Airflow 1.10.
    Those are the same providers as for 2.0 but automatically back-ported to work for Airflow 1.10. Those
    backport providers are going to be updated and released for 3 months after Apache Airflow 2.0 release.

Provider packages functionality
-------------------------------

Separate provider packages provide the possibilities that were not available in 1.10:

1. You can upgrade to latest version of particular providers without the need of Apache Airflow core upgrade.

2. You can downgrade to previous version of particular provider in case the new version introduces
   some problems, without impacting the main Apache Airflow core package.

3. You can release and upgrade/downgrade provider packages incrementally, independent from each other. This
   means that you can incrementally validate each of the provider package update in your environment,
   following the usual tests you have in your environment.

Q&A for provider packages
-------------------------

Q. When upgrading to a new Airflow version such as 2.0, but possibly 2.0.1 and beyond, is the best practice

>    to also upgrade provider packages at the same time?

Reviewer: can we convert the questions into heading so they can be linked similar to https://raw.githubusercontent.com/apache/airflow/master/docs/faq.rst

Author: I think our questions are a bit too long to make them into headers. But I divided the Q&A into sections (for now one Q/A per section) and bolded the questions there. Later we might add some more questions in each section.

   3 months after the release. We will stop releasing new updates to the backport providers afterwards.
   You will be able to continue using the provider packages that you already use and unless you need to
   get some new release of the provider that is only released for 2.0, there is no need to upgrade
   Airflow. This might happen if for example the provider is migrated to use newer version of client
   libraries or when new features/operators/hooks are added to it. Those changes will only be
   backported to 1.10.* compatible backport providers up to 3 months after releasing Airflow 2.0.
   Also we expect more providers, changes and fixes added to the existing providers to come after the
   3 months pass. Eventually you will have to upgrade to Airflow 2.0 if you would like to make use of those.
   When it comes to compatibility of providers with different Airflow 2 versions, each
   provider package will keep it's own dependencies, and while we expect those providers to be generally
   backwards-compatible, particular versions of particular providers might introduce dependencies on

>    specific Airflow versions.

Reviewer: ```suggestion
A. Backport Provider Packages (those are needed in 1.10.* Airflow series) are going to be released for
   3 months after the release. We will stop releasing new updates to the backport providers afterwards.
   You will be able to continue using the provider packages that you already use and unless you need to
   get some new release of the provider that is only released for 2.0, there is no need to upgrade
   Airflow. This might happen if for example the provider is migrated to use a newer version of client
   libraries or when new features/operators/hooks are added to it. Those changes will only be
   backported to 1.10.* compatible backport providers up to 3 months after releasing Airflow 2.0.
   Also, we expect more providers, changes and fixes added to the existing providers to come after the
   3 months pass. Eventually, you will have to upgrade to Airflow 2.0 if you would like to make use of those.
   When it comes to compatibility of providers with different Airflow 2 versions, each
   provider package will keep its own dependencies, and while we expect those providers to be generally
   backwards-compatible, particular versions of particular providers might introduce dependencies on
   specific Airflow versions.
```

Q. I have an older version of my provider package which we have lightly customized and is working
   fine with my MSSQL installation. I am upgrading my Airflow version. Do I need to upgrade my provider,
   or can I keep it as it is.

A. It depends on the scope of customization. There is no need to upgrade the provider packages to later
   versions unless you want to upgrade to Airflow version that introduces backwards incompatible changes.
   Generally speaking, with Airflow 2 we are following the `Semver <https://semver.org/>`_  approach where
   we will introduce backwards-incompatible changes in Major releases, so all your modifications (as long

>    as you have not used internal Airflow classes) should work for All Airflow 2.* versions.

Reviewer: ```suggestion
A. It depends on the scope of customization. There is no need to upgrade the provider packages to later
   versions unless you want to upgrade to Airflow version that introduces backwards-incompatible changes.
   Generally speaking, with Airflow 2 we are following the `Semver <https://semver.org/>`_  approach where
   we will introduce backwards-incompatible changes in Major releases, so all your modifications (as long
   as you have not used internal Airflow classes) should work for All Airflow 2.* versions.
```
